### ISSUE TITLE
Demo Inference script

### ISSUE THREAD

**Issue by c-lecellier** [ClaudiaCuttano/SAMWISE#22]
Hi,

Great work on SAMWISE, I wanted to try the model on my own video just for a quick demo, but I couldn’t find an easy way to do that.

I saw some issues mentioning a demo, so I made a simple script to help run the model on a custom video. I used one of the dataset inference scripts and made a quick version for demo use.

If you have any suggestions to improve it, that would be great!

Thanks again for the nice work.

Command example:
`
 python inference_demo.py --frames_dir "./path/to/frames" --expression "Your prompt" --resume "./checkpoint.pth" --output_dir "results" --visualize --HSA --use_cme_head 
`

```
'''
Inference demo code for SAMWISE, on one video
'''
from datasets.transform_utils import vis_add_mask
import argparse
import random
import time
from pathlib import Path
import numpy as np
import torch
import util.misc as utils
import os
from PIL import Image
import torch.nn.functional as F
import json
from tqdm import tqdm
from os.path import join
import sys
from tools.colormap import colormap
import opts
from models.samwise import build_samwise
from util.misc import on_load_checkpoint
from datasets.transform_utils import VideoEvalDataset
from torch.utils.data import DataLoader
import glob

# colormap
color_list = colormap()
color_list = color_list.astype('uint8').tolist()

def get_frames_from_directory(frames_dir, extension='jpg'):
    pattern = os.path.join(frames_dir, f"*.{extension}")
    frame_paths = glob.glob(pattern)
    
    # Extract frame names without extension and sort
    frame_names = []
    for path in frame_paths:
        filename = os.path.basename(path)
        frame_name = os.path.splitext(filename)[0]
        frame_names.append(frame_name)
    
    # Sort alphabetically (or numerically if they're numeric)
    frame_names.sort()
    return frame_names


def main(args):
    args.batch_size = 1
    print("Inference only supports for batch size = 1") 
    print(args)

    # save path
    output_dir = args.output_dir
    save_path_prefix = os.path.join(output_dir, 'Annotations')
    os.makedirs(save_path_prefix, exist_ok=True)
    
    args.log_file = join(args.output_dir, 'log.txt')
    with open(args.log_file, 'w') as fp:
        fp.writelines(" ".join(sys.argv)+'\n')
        fp.writelines(str(args.__dict__)+'\n\n')        

    save_visualize_path_prefix = os.path.join(output_dir, 'viz_images')
    if args.visualize:
        os.makedirs(save_visualize_path_prefix, exist_ok=True)

    # Single video and expression setup
    video_path = args.frames_dir  # Directory containing frames
    expression = args.expression  # Text expression for segmentation
    
    # Automatically get all frames from directory
    frames = get_frames_from_directory(args.frames_dir, "jpg")
    print(f"Found {len(frames)} frames in directory: {args.frames_dir}")
    print(f"Frame range: {frames[0]} to {frames[-1]}")

    start_time = time.time()
    print('Start inference')
        
    result_dict = sub_processor(args, video_path, expression, frames, 
                               save_path_prefix, save_visualize_path_prefix)

    end_time = time.time()
    total_time = end_time - start_time
    print("Total inference time: %.4f s" %(total_time))

def sub_processor(args, video_path, expression, frames, save_path_prefix, save_visualize_path_prefix):
    # model setup
    model = build_samwise(args)
    device = torch.device(args.device)
    model.to(device)

    model_without_ddp = model

    n_parameters = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print('number of params:', n_parameters)

    if args.resume:
        checkpoint = torch.load(args.resume, map_location='cpu')
        checkpoint = on_load_checkpoint(model_without_ddp, checkpoint)
        missing_keys, unexpected_keys = model_without_ddp.load_state_dict(checkpoint['model'], strict=False)
        unexpected_keys = [k for k in unexpected_keys if not (k.endswith('total_params') or k.endswith('total_ops'))]
        if len(missing_keys) > 0:
            print('Missing Keys: {}'.format(missing_keys))
        if len(unexpected_keys) > 0:
            print('Unexpected Keys: {}'.format(unexpected_keys))
    else:
        raise ValueError('Please specify the checkpoint for inference.')

    # start inference
    model.eval()
    
    # Process single video with single expression
    video_len = len(frames)
    all_pred_masks = []
    
    vd = VideoEvalDataset(video_path, frames, max_size=args.max_size)
    dl = DataLoader(vd, batch_size=args.eval_clip_window,
            num_workers=args.num_workers, shuffle=False)
    origin_w, origin_h = vd.origin_w, vd.origin_h
    
    # Process each clip
    for imgs, clip_frames_ids in dl:
        print(f"Processing clip with frames: {clip_frames_ids}")
        clip_frames_ids = clip_frames_ids.tolist()
        imgs = imgs.to(args.device)  # [eval_clip_window, 3, h, w]
        img_h, img_w = imgs.shape[-2:]
        size = torch.as_tensor([int(img_h), int(img_w)]).to(args.device)
        target = {"size": size, 'frame_ids': clip_frames_ids}

        with torch.no_grad():
            outputs = model([imgs], [expression], [target])

        pred_masks = outputs["pred_masks"]  # [t, q, h, w]
        pred_masks = pred_masks.unsqueeze(0)
        pred_masks = F.interpolate(pred_masks, size=(origin_h, origin_w), mode='bilinear', align_corners=False) 
        pred_masks = (pred_masks.sigmoid() > args.threshold)[0].cpu() 
        all_pred_masks.append(pred_masks)

    # store the video results
    all_pred_masks = torch.cat(all_pred_masks, dim=0).numpy()  # (video_len, h, w)

    # Extract video name from path for saving
    video_name = os.path.basename(video_path).split('.')[0]

    if args.visualize:
        for t, frame in enumerate(frames):
            # original image path - adjust based on your video structure
            img_path = os.path.join(video_path, frame + '.jpg')
            source_img = Image.open(img_path).convert('RGBA') # PIL image

            # draw mask
            source_img = vis_add_mask(source_img, all_pred_masks[t], color_list[0])

            # save visualization
            save_visualize_path_dir = os.path.join(save_visualize_path_prefix, video_name)
            os.makedirs(save_visualize_path_dir, exist_ok=True)
            save_visualize_path = os.path.join(save_visualize_path_dir, frame + '.png')
            source_img.save(save_visualize_path)

    # save binary masks
    save_path = os.path.join(save_path_prefix, video_name)
    os.makedirs(save_path, exist_ok=True)
    for j in range(video_len):
        frame_name = frames[j]
        mask = all_pred_masks[j].astype(np.float32) 
        mask = Image.fromarray(mask * 255).convert('L')
        save_file = os.path.join(save_path, frame_name + ".png")
        mask.save(save_file)

    return {"frames_processed": video_len}
    

# visuaize functions
def box_cxcywh_to_xyxy(x):
    x_c, y_c, w, h = x.unbind(1)
    b = [(x_c - 0.5 * w), (y_c - 0.5 * h),
            (x_c + 0.5 * w), (y_c + 0.5 * h)]
    return torch.stack(b, dim=1)

def rescale_bboxes(out_bbox, size):
    img_w, img_h = size
    b = box_cxcywh_to_xyxy(out_bbox)
    b = b.cpu() * torch.tensor([img_w, img_h, img_w, img_h], dtype=torch.float32)
    return b


# Visualization functions
def draw_reference_points(draw, reference_points, img_size, color):
    W, H = img_size
    for i, ref_point in enumerate(reference_points):
        init_x, init_y = ref_point
        x, y = W * init_x, H * init_y
        cur_color = color
        draw.line((x-10, y, x+10, y), tuple(cur_color), width=4)
        draw.line((x, y-10, x, y+10), tuple(cur_color), width=4)

def draw_sample_points(draw, sample_points, img_size, color_list):
    alpha = 255
    for i, samples in enumerate(sample_points):
        for sample in samples:
            x, y = sample
            cur_color = color_list[i % len(color_list)][::-1]
            cur_color += [alpha]
            draw.ellipse((x-2, y-2, x+2, y+2), 
                            fill=tuple(cur_color), outline=tuple(cur_color), width=1)

def vis_add_mask(img, mask, color):
    origin_img = np.asarray(img.convert('RGB')).copy()
    color = np.array(color)

    mask = mask.reshape(mask.shape[0], mask.shape[1]).astype('uint8') # np
    mask = mask > 0.5

    origin_img[mask] = origin_img[mask] * 0.5 + color * 0.5
    origin_img = Image.fromarray(origin_img)
    return origin_img



if __name__ == '__main__':
    if torch.cuda.get_device_properties(0).major >= 8:
        torch.backends.cuda.matmul.allow_tf32 = True
        torch.backends.cudnn.allow_tf32 = True

    parser = argparse.ArgumentParser('SAMWISE Inference demo one video script', parents=[opts.get_args_parser()])
    # Command example:
    # python inference_demo.py --frames_dir "./path/to/frames" --expression "Your Prompt" --resume "./checkpoint.pth" --output_dir "results" --visualize --HSA --use_cme_head


    # Updated arguments for directory-based frame processing
    parser.add_argument('--frames_dir', type=str, required=True, help='Path to directory containing video frames')
    parser.add_argument('--expression', type=str, required=True, help='Text expression for object segmentation')
    
    args = parser.parse_args()
    args.output_dir = os.path.join(args.output_dir, "demo")

    main(args)

**Comment by c-lecellier** [comment]
> thanks for your share ! have you try to inference on multi-gpus?

Unfortunately, I couldn't try it because I don't have the necessary hardware.

**Comment by SaharGezer** [comment]
> > Hey which weights did you use as --resume "./checkpoint.pth"? [@c-lecellier](https://github.com/c-lecellier)
> 
> Hey!
> 
> You can try with these weights here: https://github.com/ClaudiaCuttano/SAMWISE?tab=readme-ov-file#-referring-video-object-segmentation-rvos

Thank you very much, it works

**Comment by SaharGezer** [comment]
@c-lecellier Does using a batch size of 1 affect the model’s memory usage between frames?

**Comment by ga1i13o** [comment]
Hi @c-lecellier,
apologies for the late reply.

Thanks for your message and for sharing the script.

We've just updated the repo with an `inference_demo.py` script to the repository that supports running SAMWISE on custom inputs (images or videos). It's documented in  [README: Try SAMWISE on Your Own Data](https://github.com/ClaudiaCuttano/SAMWISE#%EF%B8%8F-try-samwise-on-your-own-data).

You're welcome to open a pull request if you'd like to extend or improve the demo further, contributions are always appreciated!

Thanks again.
